### PR TITLE
ONT-707 optimize vbft make proposal

### PR DESCRIPTION
**Pull request by xiemylogos**

### CODE REVIEW

File: consensus/vbft/service.go [hunk 1]
```diff
@@ -1058,11 +1048,18 @@ func (self *Server) onConsensusMsg(peerIdx uint32, msg ConsensusMsg, msgHash com
...
-	blk, _ := self.blockPool.getSealedBlock(msg.GetBlockNum() - 1)
+	blk, prevBlkHash := self.blockPool.getSealedBlock(msg.GetBlockNum() - 1)
 	if blk == nil {
 		log.Errorf("BlockProposal failed to GetPreBlock:%d", (msg.GetBlockNum() - 1))
 		return
 	}
+
+	msgPrevBlkHash := msg.Block.getPrevBlockHash()
+	if bytes.Compare(prevBlkHash.ToArray(), msgPrevBlkHash.ToArray()) != 0 {
+		log.Errorf("BlockPrposalMessage check blocknum:%d,prevhash:%d,msg prevhash:", msg.GetBlockNum(), prevBlkHash, msgPrevBlkHash)
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go]: print hash with hex.encode

File: core/store/ledgerstore/ledger_store.go [hunk 2]
```diff
@@ -25,6 +25,8 @@ import (
 	"strings"
 	"sync"
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go]: go import fmt

File: core/store/ledgerstore/ledger_store.go [hunk 3]
```diff
@@ -387,12 +388,12 @@ func (this *LedgerStoreImp) verifyHeader(header *types.Header) error {
...
 	}
 	consensusType := strings.ToLower(config.DefConfig.Genesis.ConsensusType)
-	if consensusType != "vbft" {
 
-		if prevHeader.Timestamp >= header.Timestamp {
-			return fmt.Errorf("block timestamp is incorrect")
-		}
+	if prevHeader.Timestamp >= header.Timestamp {
+		return fmt.Errorf("block timestamp is incorrect")
+	}
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go]: move 'consensusType := strings.ToLower(config.DefConfig.Genesis.ConsensusType)'  here.

**Review by Honglei-Cong — CHANGES_REQUESTED** [review]

**Review by Honglei-Cong — DISMISSED** [review]

File: consensus/vbft/service.go [hunk 4]
```diff
@@ -1058,11 +1048,18 @@ func (self *Server) onConsensusMsg(peerIdx uint32, msg ConsensusMsg, msgHash com
...
 	msgBlkNum := msg.GetBlockNum()
-	blk, _ := self.blockPool.getSealedBlock(msg.GetBlockNum() - 1)
+	blk, prevBlkHash := self.blockPool.getSealedBlock(msg.GetBlockNum() - 1)
 	if blk == nil {
 		log.Errorf("BlockProposal failed to GetPreBlock:%d", (msg.GetBlockNum() - 1))
 		return
 	}
+
+	msgPrevBlkHash := msg.Block.getPrevBlockHash()
+	if bytes.Compare(prevBlkHash.ToArray(), msgPrevBlkHash.ToArray()) != 0 {
```
Comment (laizy) [inline consensus/vbft/service.go]: you can use `prevBlkHash != msgPrevBlkHash` directly

**Review by laizy — CHANGES_REQUESTED** [review]

File: consensus/vbft/service.go [hunk 5]
```diff
@@ -626,14 +625,15 @@ func (self *Server) getState() ServerState {
...
 	blkNum := self.GetCurrentBlockNo()
-
 	block, _ := self.blockPool.getSealedBlock(blkNum - 1)
 	if block == nil {
 		return fmt.Errorf("failed to get sealed block (%d)", blkNum-1)
 	}
-
 	var err error
 	self.metaLock.Lock()
+	if self.checkNeedUpdateChainConfig(blkNum) || self.checkUpdateChainConfig() {
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go]: do this check before metaLock.Lock() ?

Comment (Honglei-Cong) [reply consensus/vbft/service.go]: btw, 
always self.config = block.Info.NewChainConfig when block.Info.NewChainConfig is not nil ?

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**